x-pack/plugins/canvas/canvas_plugin_src/plugin.ts

```diff
@@ -37,9 +40,10 @@ export class CanvasSrcPlugin implements Plugin<void, void, SetupDeps, StartDeps>
 
     plugins.canvas.addRenderers(renderFunctions);
 
-    core.getStartServices().then(([coreStart, depsStart]) => {
+    core.getStartServices().then(([coreStart, startPlugins]) => {
+      plugins.canvas.addFunctionFactories(factories, { coreStart, startPlugins });
```

Comment:
This abstraction feels to general to me and not general enough on the other hand.

We are giving whole coreStart and all plugins start contracts to each factory. Most factories need a single thing. And possibly it could be something that is not provided on any start contract (but something internal to the plugin providing the function). When you will try to use this code on the server you will have a lot of problems as server and client side start contracts don't look the same.

I think we should rather call each factory here and provide it just what it needs.

```
  plugins.canvas.registerFunction(esdocsFactory({ search: startPlugins.data.search) });
```

this will make it clearer what the dependencies are, it will make it easier to make code available on the server (you will only need to abstract search behind a common server/client interface) and it gives flexibility for whatever random thing you need to provide to some other functions in the future.